blogposts/2020/find-and-replace.md

```diff
@@ -0,0 +1,582 @@
+
+This post presents a guide to all the various find-and-replace tools I've encountered and used over
+the years, divided into three chapters:
+
+1. [Find and replace in your editor](#find-and-replace-in-your-editor)
+   * [Regular expressions](#regular-expressions)
+   * [Keyboard macros and multiselect](#keyboard-macros-and-multiselect)
+   * [Semantic refactoring](#semantic-refactoring)
+1. [Find and replace outside your editor, on your local machine](#find-and-replace-outside-your-editor)
+   * [`grep` and `sed`](#grep-and-sed)
+   * [Codemod](#codemod)
+   * [Language-specific tools](#language-specific-tools)
+   * [Comby](#comby)
+1. [Making large-scale code modifications tractable](#beyond-your-local-machine)
+   * [Sourcegraph Campaigns](#sourcegraph-campaigns)
+
+If you're already familiar with some of these tools, feel free to skip to the sections that you're
+unfamiliar with. If you're new to all this, my hope is that this post will give you a lay of the
+land and serve as a jumping-off point to learn about a range of powerful tools that will save you a
+lot of time and aggravation over your programming life.
+
+
+## Find and replace in your editor
+
+Most code editors offer some sort of find-and-replace facility. At the most basic level, you have
+literal string substitution. This enables you to find instances of `foo` and replace them with
+`bar`.
+
+However, there are many cases where you want to apply a general change pattern, not just replace one
+word with another. Adding arguments to a function, fixing a common anti-pattern, standardizing
+library usage, and transforming data from one format to another—these call for a find-and-replace
+tool that can express patterns of transformation.
+
+### Regular expressions
+
+The most commonly used pattern-matching language is Regular Expressions, commonly abbreviated as
+"regex".[^1] Most code editors support regex searching, and it is commonly toggled on using a button
+or icon with the `.*` symbol.
+
+With regexes, you can do things like this:
+
+| Description                                         | Regex      | Match              | Does not match |
+|-----------------------------------------------------|------------|--------------------|----------------|
+| Find all symbol names starting with "http"          | `foo\w*`   | fooBar             | barFoo         |
+| Find all characters between double quotes           | `"[^"]"`   | "hello&nbsp;world" | hello world'   |
+| Find all references to fields of a certain variable | `\w+\.\w+` | base.Path          | basePath       |
+
+There are different dialects or flavors of regex. In this post, we'll use [**POSIX Extended Regular
+Expressions**](https://www.gnu.org/software/findutils/manual/html_node/find_html/posix_002dextended-regular-expression-syntax.html#posix_002dextended-regular-expression-syntax). Here's
+a short primer on its syntax:
+
+* Alphanumeric characters are generally interpreted literally.
```

Comment:
Feels too elementary/reference-like to be showing this here. Who's the audience? What's the goal?

I think the audience should be senior engineers and the goal should be to make people realize "hmm, yeah, i as a dev have made the progression from find in my editor, to grep, to codemod...and there is a gap in my needs that something like campaigns fills (or the next step in my evolution is something like campaigns)...oh, and google's been using something like it for a while too?" (and check out what fatih mentions at https://twitter.com/fatih/status/1259912881186824192 and cite that too)